pkg/gds/admin.go

```diff
@@ -1139,6 +1149,89 @@ func (s *Admin) updateVASPTRIXO(vasp *pb.VASP, data map[string]interface{}, log
 	return true, http.StatusOK, nil
 }
 
+// Update the VASP contacts; this completely overwrites all of the previous contact data.
+func (s *Admin) updateVASPContacts(vasp *pb.VASP, data map[string]interface{}, log zerolog.Logger) (_ bool, _ int, err error) {
+	var errs *multierror.Error
+
+	// Check if contact data has been supplied, otherwise do not update.
+	if len(data) == 0 {
+		return false, http.StatusOK, nil
+	}
+
+	// Remarshal the JSON contact data.
+	contacts := &pb.Contacts{}
+	if err = wire.Unwire(data, contacts); err != nil {
+		log.Warn().Err(err).Msg("could not unwire JSON data into a valid Contacts")
+		return false, http.StatusBadRequest, errors.New("could not parse Contacts")
+	}
+
+	// Iterate through the existing VASP contacts, copying the data fields from the
+	// provided contacts without overwriting the extra data.
+	iter := models.NewContactIterator(vasp.Contacts, false, false)
+	for iter.Next() {
+		old, kind := iter.Value()
+		new := models.ContactFromType(contacts, kind)
+		if new != nil {
+			old.Name = new.Name
+			old.Phone = new.Phone
+			old.Person = new.Person
+
+			if new.Email != "" && new.Email != old.Email {
+				// The email address changed, so the contact needs to be verified.
+				old.Email = new.Email
+
+				if err := models.SetContactVerification(old, secrets.CreateToken(models.VerificationTokenLength), false); err != nil {
+					errs = multierror.Append(errs, err)
+					continue
+				}
+
+				// Only send an email to the updated contact.
+				if err := s.svc.email.SendVerifyContact(vasp, old); err != nil {
+					errs = multierror.Append(errs, err)
+					continue
+				}
+
+				// Write the updated email log to the database.
+				if err := s.db.UpdateVASP(vasp); err != nil {
+					errs = multierror.Append(errs, err)
+					continue
+				}
+			}
+		} else {
+			// Need to use the VASP pointer to remove the contact.
+			models.DeleteContact(vasp, kind)
+		}
+	}
+
+	// Iterate through the provided contacts, adding any new contacts that don't exist
+	// on the VASP.
+	iter = models.NewContactIterator(contacts, false, false)
+	for iter.Next() {
+		new, kind := iter.Value()
+		old := models.ContactFromType(vasp.Contacts, kind)
+		if old == nil {
+			// Need to use the VASP pointer to add the contact.
+			if err := models.AddContact(vasp, kind, new); err != nil {
+				errs = multierror.Append(errs, err)
+			}
+
+			// Send a verification email to the new contact.
+			if err := s.svc.email.SendVerifyContact(vasp, new); err != nil {
+				errs = multierror.Append(errs, err)
+				continue
+			}
+
+			// Write the updated email log to the database.
+			if err := s.db.UpdateVASP(vasp); err != nil {
+				errs = multierror.Append(errs, err)
+				continue
+			}
+		}
+	}
```

Comment:
This code could be simplified a lot if we don't allow contacts to be added/removed on the update. Let me know what you think.